src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+                lock (this) {
+                    foreach (var disposeScope in DisposeScopeStack) {
+                        size += disposeScope.TotalAllocatedSize;
+                    }
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            Interlocked.Increment(ref _createdCount);
+
+            lock (this) {
+                if (DisposeScopeStack.Count > 0) {
+                    DisposeScopeStack.Peek().Include(disposable);
+                }
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            Interlocked.Increment(ref _disposedCount);
+            lock (this) {
+                foreach (var disposeScope in DisposeScopeStack) {
+                    disposeScope.Disposables.Remove(disposable);
+                }
+            }
+        }
+
+        internal static DisposeScope NewDisposeScope()
+        {
+            lock (Singleton) {
+                return Singleton.InnerNewDisposeScope();
+            }
+        }
+
+        private DisposeScope InnerNewDisposeScope()
+        {
+            lock (this) {
+                var disposeScope = new DisposeScope(this);
+                DisposeScopeStack.Push(disposeScope);
+                return disposeScope;
+            }
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                Debug.Assert(DisposeScopeStack.Count > 0);
+                Debug.Assert(DisposeScopeStack.Peek() == disposeScope);
+                DisposeScopeStack.Pop();
+            }
+        }
+
+        internal void MoveToOuterScope(DisposeScope disposeScope, IDisposable disposable)
+        {
+            lock (this) {
+                var array = DisposeScopeStack.ToArray();
+                for (var i = array.Length - 1; i >= 1; i--) {
+                    if (array[i] == disposeScope) {
+                        array[i - 1].Include(disposable);
+                        return;
+                    }
+                }
+            }
+        }
+
+        public class DisposeScope : IDisposable
+        {
+            private readonly DisposeScopeManager _disposeScopeManager;
+
+            public DisposeScope(DisposeScopeManager disposeScopeManager)
+            {
+                _disposeScopeManager = disposeScopeManager;
+            }
+
+            /// <summary>
+            /// The disposables that are scheduled for disposing.
+            /// </summary>
+            public HashSet<IDisposable> Disposables { get; } = new();
+
+            public long TotalAllocatedSize =>
+                Disposables.OfType<torch.Tensor>().Sum(x => x.NumberOfElements * x.ElementSize);
```

Comment:
This is misleading, as multiple tensors can share the same memory.

Reply:
I've removed it, though I wish there was a way to monitor memory usage.